p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,465 @@
+
+const (
+	// hansdhakeChannelNegotiated is used to specify that the
+	// handshake data channel does not need negotiation via DCEP.
+	// A constant is used since the `DataChannelInit` struct takes
+	// references instead of values.
+	hansdhakeChannelNegotiated = true
+	// handshakeChannelId is the agreed ID for the handshake data
+	// channel.
+	// A constant is used since the `DataChannelInit` struct takes
+	// references instead of values. We specify the type here as this
+	// value is only ever copied and passed by reference
+	handshakeChannelId = uint16(0)
+)
+
+// timeout values for the peerconnection
+// https://github.com/pion/webrtc/blob/v3.1.50/settingengine.go#L102-L109
+const DefaultDisconnectedTimeout = 5 * time.Second
+const DefaultFailedTimeout = 25 * time.Second
+const DefaultKeepaliveTimeout = 2 * time.Second
+
+type WebRTCTransport struct {
+	webrtcConfig webrtc.Configuration
+	rcmgr        network.ResourceManager
+	privKey      ic.PrivKey
+	noiseTpt     *noise.Transport
+	localPeerId  peer.ID
+
+	// timeouts
+	peerConnectionFailedTimeout       time.Duration
+	peerConnectionDisconnectedTimeout time.Duration
+	peerConnectionKeepaliveTimeout    time.Duration
+
+	// in-flight connections
+	maxInFlightConnections uint32
+}
+
+var _ tpt.Transport = &WebRTCTransport{}
+
+type Option func(*WebRTCTransport) error
+
+// WithPeerConnectionIceTimeouts sets the ice disconnect, failure and keepalive timeouts
+func WithPeerConnectionIceTimeouts(disconnect time.Duration, failed time.Duration, keepalive time.Duration) Option {
+	return func(t *WebRTCTransport) error {
+		if failed < disconnect {
+			return fmt.Errorf("disconnect timeout cannot be greater than failed timeout")
+		}
+		if disconnect <= keepalive {
+			return fmt.Errorf("keepalive timeout cannot be greater than or equal to failed timeout")
+		}
+		t.peerConnectionDisconnectedTimeout = disconnect
+		t.peerConnectionFailedTimeout = failed
+		t.peerConnectionKeepaliveTimeout = keepalive
+		return nil
+	}
+}
+
+// WithListenerMaxInFlightConnections sets the maximum number of connections that are in-flight, i.e
+// they are being negotiated, or are waiting to be accepted.
+func WithListenerMaxInFlightConnections(m uint32) Option {
+	return func(t *WebRTCTransport) error {
+		t.maxInFlightConnections = m
+		return nil
+	}
+}
+
+func New(privKey ic.PrivKey, psk pnet.PSK, gater connmgr.ConnectionGater, rcmgr network.ResourceManager, opts ...Option) (*WebRTCTransport, error) {
+	if psk != nil {
+		log.Error("WebRTC doesn't support private networks yet.")
+		return nil, fmt.Errorf("WebRTC doesn't support private networks yet")
+	}
+	localPeerId, err := peer.IDFromPrivateKey(privKey)
+	if err != nil {
+		return nil, fmt.Errorf("could not get local peer ID: %w", err)
+	}
+	// We use elliptic P-256 since it is widely supported by browsers.
+	// See: https://github.com/libp2p/specs/pull/412#discussion_r968294244
```

Comment:
the link above does not really tell anything to an outsider? Perhaps link to an actual spec paragraph?!

Reply:
fixed in https://github.com/libp2p/go-libp2p/pull/1999/commits/f76dfc499857bffe287bedd931eeba6c652a4825 (first commit of new PR https://github.com/libp2p/go-libp2p/pull/1999)